File: tests/bench/05_maps.go
// Benchmark 05: Maps — build, lookup, membership
// Measures map insertion, O(1) lookup, membership check
package main

import (
	"fmt"
	"time"
)

func main() {
	// Part A: Build map
	nBuild := 500_000
	start := time.Now()
	m := make(map[string]int64)
	for i := int64(0); i < int64(nBuild); i++ {
		m[fmt.Sprintf("k%d", i)] = i
	}
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("[bench] map_build: %d entries in %d ms\n", nBuild, elapsed)

	// Part B: Lookup every key
	start = time.Now()
	total := int64(0)
	for i := int64(0); i < int64(nBuild); i++ {
		total += m[fmt.Sprintf("k%d", i)]
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] map_lookup: %d lookups in %d ms (sum=%d)\n", nBuild, elapsed, total)

	// Part C: Membership check
	nHas := 1_000_000
	start = time.Now()
	found := int64(0)
	for i := int64(0); i < int64(nHas); i++ {
		k := fmt.Sprintf("k%d", i)
		if _, ok := m[k]; ok {
			found++
		}
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] map_has: %d checks in %d ms (found=%d)\n", nHas, elapsed, found)
}

File: tests/bench/05_maps.rs
// Benchmark 05: Maps — build, lookup, membership
// Measures HashMap insertion, O(1) lookup, contains_key

use std::collections::HashMap;
use std::time::Instant;

fn main() {
    // Part A: Build map
    let n_build = 500_000;
    let start = Instant::now();
    let mut m: HashMap<String, i64> = HashMap::new();
    for i in 0..n_build {
        m.insert(format!("k{}", i), i);
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] map_build: {} entries in {} ms", n_build, elapsed);

    // Part B: Lookup every key
    let start = Instant::now();
    let mut total: i64 = 0;
    for i in 0..n_build {
        total += m[&format!("k{}", i)];
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] map_lookup: {} lookups in {} ms (sum={})", n_build, elapsed, total);

    // Part C: contains_key membership
    let n_has = 1_000_000;
    let start = Instant::now();
    let mut found: i64 = 0;
    for i in 0..n_has {
        let k = format!("k{}", i);
        if m.contains_key(&k) {
            found += 1;
        }
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] map_has: {} checks in {} ms (found={})", n_has, elapsed, found);
}
